Singapore Investor Alert List

nvmbrasserie opened the pull request:
fixes opensanctions/crawler-planning#317

nvmbrasserie committed:
initial commit

nvmbrasserie committed:
fully mapped

nvmbrasserie committed:
email lookups

nvmbrasserie committed:
all warnings handled

nvmbrasserie committed:
assertions added

nvmbrasserie committed:
Merge branch 'main' into sg-mas-investor-alert-list

nvmbrasserie committed:
topics added

nvmbrasserie committed:
splits for names, aliases and websites

nvmbrasserie committed:
fixed lookups

nvmbrasserie committed:
names and websites mapped

nvmbrasserie committed:
owenrships mapped

nvmbrasserie committed:
ownerships mapped

nvmbrasserie committed:
websites cleaned up

nvmbrasserie committed:
restructured

nvmbrasserie committed:
even cleaner names

leonhandreke reviewed (commented):
Cool, this looks like a nice data source and a clean crawler! If only they had used their own website field like it was meant to be used...

nvmbrasserie commented on datasets/sg/mas_investor_alert/crawler.py:
```diff
@@ -0,0 +1,116 @@
+        context.data_url,
+        headers=HEADERS,
+    )
+    context.export_resource(path, JSON, title=context.SOURCE_TITLE)
```
I think that when a source provides structured data in a single request (like a complete JSON response), it's nice to export it entirely.  But when the data is scattered across multiple pages (like Bulgarian declarations case), then we don't want to export them all

leonhandreke replied:
Thanks for clarifying, today I learned :)

nvmbrasserie commented on datasets/sg/mas_investor_alert/crawler.py:
```diff
@@ -0,0 +1,116 @@
+    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/18.2 Safari/605.1.15",
+}
+IGNORE = [
+    "phonenumber_s",
```
yeah, but for some sources it's considered an overkill (peps is a good example here). I think it's the same for this source here

leonhandreke replied:
Looks like they're all companies, right? Let's add them and give the crypto scammers a call :P

nvmbrasserie committed:
metadata post-review

nvmbrasserie committed:
more post review

nvmbrasserie committed:
cleaner way to assign ids

nvmbrasserie committed:
Add detection for missing entity links. Now we track which entities exist and which are referenced, so we can
identify broken links at the end of the crawl.

nvmbrasserie commented on datasets/sg/mas_investor_alert/crawler.py:
```diff
@@ -0,0 +1,143 @@
+WEBSITE_KEYWORDS = [".com", ".net", ".org", "https:", "http:", "www.", ".sq", ".co"]
+NON_EXISTENT_RELATIONS = {"5685", "4891"}
+
+all_entities = set()
+all_related = set()
```
this is not how we usually do this, but I think it's much cleaner way here than passing it as an argument starting from `crawl()`

leonhandreke replied:
I see the issue that for every dangling relationship, we'll emit broken data and generate a warning, but we can't really do anything but acknowledge it, right? How do you feel about https://github.com/opensanctions/opensanctions/pull/2111/files

nvmbrasserie replied:
@jbothma said it works. Let's go ahead and merge it

nvmbrasserie committed:
phones added

leonhandreke committed:
[sg_mas_investor_alert] Never emit broken relationships (#2111)

* [sg_mas_investor_alert] Never emit broken relationships

In the previous version, we would still emit them until somebody saw the
warning, came in, added to the list. But we can't really do anything
else, and if they show up later, we'll still skip them. So do this
instead, where we remember what source IDs we've seen, and don't emit
the relation if we didn't see it.

leonhandreke reviewed (approved):
🚀
